lib/active_hash/base.rb

```diff
@@ -136,19 +136,23 @@ def exists?(record)
 
       def insert(record)
         @records ||= []
-        @max_id ||= 0
-        if record[:id] && record[:id].is_a?(Numeric)
-          @max_id = [@max_id, record[:id]].max
-        else
-          record[:id] ||= (@max_id = @max_id.succ)
-        end
+        set_id(record)
         validate_unique_id(record) if dirty
         mark_dirty
 
         add_to_record_index({ record.id.to_s => @records.length })
         @records << record
       end
 
+      def set_id(record)
+        # sets record[:id] to @max_id+1 if it doesn't exist
+        if record[:id] && record[:id].is_a?(Numeric)
+          @max_id = [@max_id || 0, record[:id].ceil].max
+        else
+          record[:id] ||= (@max_id = @max_id.to_i.succ)
+        end
```

Comment:
I didn't initialise `@max_id` to handle the negative values but if you are okay with ignoring the negative IDs, then you can be more expressive (similar to your original PR) -- no need for hiding `nil.to_i = 0` in `@max_id.to_i.succ` and  `[@max_id || 0, ...].max`.

```ruby
      def set_id(record)
        # sets record[:id] to @max_id+1 if it doesn't exist
        @max_id ||= 0
        if record[:id] && record[:id].is_a?(Numeric)
          @max_id = [@max_id, record[:id].ceil].max
        else
          record[:id] ||= (@max_id = @max_id.succ)
        end
      end
```